crates/apub/src/http/inbox.rs

```diff
@@ -0,0 +1,171 @@
+use lemmy_api_common::context::LemmyContext;
+use lemmy_utils::error::LemmyResult;
+use once_cell::sync::Lazy;
+use std::{
+  cmp::Ordering,
+  collections::BinaryHeap,
+  sync::{Arc, RwLock},
+  thread::available_parallelism,
+  time::Duration,
+};
+use tokio::{spawn, task::JoinHandle, time::sleep};
+use tracing::info;
+
+/// Handle incoming activities.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  bytes: Bytes,
+  data: Data<LemmyContext>,
+) -> LemmyResult<HttpResponse> {
+  match serde_json::from_slice::<WithPublished<()>>(&bytes)?.published {
+    Some(published) => {
+      // includes published timestamp, insert to queue to ensure that activities are processed
+      // in correct order even when delivered out of order.
+      let request_parts = (
+        request.headers().clone(),
+        request.method().clone(),
+        request.uri().clone(),
+      );
+      ACTIVITY_QUEUE.write().unwrap().push(InboxActivity {
+        request_parts,
+        bytes,
+        published,
+      });
+      Ok(HttpResponse::Ok().finish())
+    }
+    None => {
+      // no timestamp included, process immediately
+      receive_activity::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+        request, bytes, &data,
+      )
+      .await
+    }
+  }
+}
+
+/// Queue of incoming activities, ordered by oldest published first
+static ACTIVITY_QUEUE: Lazy<Arc<RwLock<BinaryHeap<InboxActivity>>>> =
+  Lazy::new(|| Arc::new(RwLock::new(BinaryHeap::new())));
+
+/// Minimum age of an activity before it gets processed. This ensures that an activity which was
+/// delayed still gets processed in correct order.
+const RECEIVE_DELAY: Option<TimeDelta> = TimeDelta::try_seconds(1);
+
+pub fn handle_received_activities(
+  context: &Data<LemmyContext>,
+) -> LemmyResult<Vec<JoinHandle<()>>> {
+  // launch one task per cpu core
+  let parallelism = available_parallelism()?.into();
+  let workers = (0..parallelism)
+    .map(|_| {
+      let context = context.reset_request_count();
+      spawn(async move {
+        loop {
+          let now = Local::now();
+          if let Some(latest_timestamp) = peek_queue_timestamp() {
+            if latest_timestamp < now - RECEIVE_DELAY.unwrap() {
+              if let Some(a) = pop_queue() {
+                let parts = (&a.request_parts.0, &a.request_parts.1, &a.request_parts.2);
+                receive_activity_parts::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+                  parts, a.bytes, &context,
+                )
+                .await
+                .inspect_err(|e| info!("Error receiving activity: {e}"))
+                .ok();
+              }
+            }
+          }
+          // TODO: could sleep based on remaining time until head activity reaches 1s
+          //       or simply use `WORK_FINISHED_RECHECK_DELAY` from lemmy_federate
+          sleep(Duration::from_millis(100)).await;
+          // TODO: need cancel? lemmy seems to shutdown just fine
```

Comment:
Lemmy shutdown works fine because you're using `tokio::spawn`. Tokio spawn is nice but when you stop the process (with any signal), all spawned tasks are immediately stopped polling - this means they will suddenly stop at any await point. This is not good for this queue because then every event still in the queue is essentially lost and any in-flight handling will be broken. 

I think this needs to use a non-spawn method of starting the tasks. It's a bit more complex... Because we need to make sure (a) no new items are added to the queue during shutdown and (b) the queue is emptied before the process exits (since it's in memory). You should be able to use the CancellableTask I implemented for the federation. Use that outside of the `loop` (not instead of it), and mostly ignore the cancellation token you receive. We don't want to cancel this task except when the queue is empty. That way the task will only allow exiting the process when the queue is fully emptied.

Here's some pseudo code
```rust
CancellableTask.spawn(|cancel| {
  // this lambda only exits when the queue is completely empty.
  while (queue not empty) {
    if (oldest queue item less than a second ago) {
      sleep (100); continue;
    }
    take item and process it.
  }
  if cancel.cancelled { return ao}
  else sleep (100) // need to sleep because the task is restarted immediately otherwise
})
```

Then, in addition, in the queue push handler (shared_inbox function) You need to if the process is shutting down reject the HTTP request with 50x. For example by having the binary heap be optional and erasing it once the process is in the shutdown phase - actually I guess this is  similar to that ActivityChannel::retrieve_activity thing we have? There's an optional and shutdown() method there but it's only used for local activities. Maybe some overlap can be used as an advantage.